zebra-test/src/service_extensions.rs

```diff
@@ -0,0 +1,33 @@
+//! Extension traits for [`Service`] types to help with testing.
+
+use futures::future::{BoxFuture, FutureExt};
+use tower::{Service, ServiceExt};
+
+use now_or_later::NowOrLater;
+
+/// An extension trait to check if a [`Service`] is immediately ready to be called.
+pub trait IsReady<Request>: Service<Request> {
+    /// Check if the [`Service`] is immediately ready to be called.
+    fn is_ready(&mut self) -> BoxFuture<bool>;
+
+    /// Check if the [`Service`] is not immediately ready because it returns an error.
+    fn not_ready_due_to_error(&mut self) -> BoxFuture<bool>;
+}
+
+impl<S, Request> IsReady<Request> for S
+where
+    S: Service<Request> + Send,
+    Request: 'static,
+{
+    fn is_ready(&mut self) -> BoxFuture<bool> {
+        NowOrLater(self.ready())
+            .map(|ready_result| matches!(ready_result, Some(Ok(_))))
+            .boxed()
+    }
+
+    fn not_ready_due_to_error(&mut self) -> BoxFuture<bool> {
```

Comment:
Naming: Can we try to avoid negated names, particularly for boolean functions?

This is consistent with the `Service::poll_ready` contract:
> If Poll::Ready(Err(_)) is returned, the service is no longer able to service requests and the caller should discard the service instance.

https://docs.rs/tower-service/0.3.1/tower_service/trait.Service.html

```suggestion
    fn is_failed(&mut self) -> BoxFuture<bool> {
```